go-metadata/istio.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istiometadata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+)
+
+// ConnectionMetadata returns the metadata about a connection
+type ConnectionMetadata struct {
+	// Identity provides the identity of the peer.
+	// Example: spiffe://cluster.local/ns/a/sa/b.
+	Identity string `json:"identity"`
+}
+
+// metadataClient provides a default client for all metadata requests
+var metadataClient = http.Client{
+	Timeout: time.Second,
+}
+
+// metadataContextKey is the key the metadata handler will store connection metadata in
+var metadataContextKey = &contextKey{}
+
+type contextKey struct {
+	name string
+}
+
+func (k *contextKey) String() string { return "istio metadata context value" }
+
+const mdsHostEnv = "GCE_METADATA_HOST"
+
+// For now, we use a well-known IP which is intercepted.
+// TODO: consider creating a real service with node-affinity to drop redirection dependency.
+const defaultHost = "169.254.169.111"
+
+func metadataServerURL() *url.URL {
+	host := os.Getenv(mdsHostEnv)
+	if host == "" {
+		host = defaultHost
+	}
+	u, err := url.Parse("http://" + host)
+	if err != nil {
+		panic(err.Error())
+	}
+	return u
+}
+
+func lookup(src, dst string) (*ConnectionMetadata, error) {
+	u := metadataServerURL()
+	params := url.Values{}
+	params.Add("src", src)
+	params.Add("dst", dst)
```

Comment:
Getting connection metadata is very useful - as Louis mentioned we may extend this to return all the details
about the connection, including labels and more handshake info. 

And since we return it as json - we may as well consider an MDS endpoint that is also signing it.

But I think the use case for passing both src and dst is primarily for dealing with CNI implementations 
where the source IP can't be preserved. Calico for example seems to support that, and hopefully we can
do it for more CNIs.  

In the case of using ztunnel in a 'best practice' CNI - we may want to support metadata lookup using only the source IP ( which would return data from the PTR-DS/WDS ). 
If the ztunnel doesn't preserve the source IP - I think the lookup result should include the original source IP 
in addition to the peer identity. And maybe all the metadata lookup info too.